src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,370 @@
+			while( extracted != null && remainingSize > 0 );
+
+			// Done?
+			if( remainingSize <= 0 )
+			{
+				break;
+			}
+		}
+
+		if( gathered != null )
+		{
+			if( mode == Actionable.MODULATE )
+			{
+				this.onTick();
+			}
+
+			return AEItemStack.create( gathered );
+		}
+
+		return null;
+
+	}
+
+	public TickRateModulation onTick()
+	{
+		LinkedList<IAEItemStack> changes = new LinkedList<IAEItemStack>();
+
+		int slots = itemHandler.getSlots();
+
+		// Make room for new slots
+		if ( slots > cachedStacks.length ) {
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		for( int slot = 0; slot < slots; slot++ )
+		{
+			// Save the old stuff
+			ItemStack oldIS = cachedStacks[slot];
+			IAEItemStack oldAeIS = cachedAeStacks[slot];
+
+			ItemStack newIS = itemHandler.getStackInSlot( slot );
+
+			if( this.isDifferent( newIS, oldIS ) )
+			{
+				// Completely different item
+				cachedStacks[slot] = newIS;
+				cachedAeStacks[slot] = AEItemStack.create( newIS );
+
+				// If we had a stack previously in this slot, notify the newtork about its disappearance
+				if( oldAeIS != null )
+				{
+					oldAeIS.setStackSize( -oldAeIS.getStackSize() );
+					changes.add( oldAeIS );
+				}
+
+				// Notify the network about the new stack. Note that this is null if newIS was null
+				if( cachedAeStacks[slot] != null )
+				{
+					changes.add( cachedAeStacks[slot] );
+				}
+			}
+			else
+			{
+				// Still the same item, but amount might have changed
+				int newSize = ( newIS == null ? 0 : newIS.stackSize );
+				int diff = newSize - ( oldIS == null ? 0 : oldIS.stackSize );
+
+				IAEItemStack stack = ( oldAeIS == null ? AEItemStack.create( newIS ) : oldAeIS.copy() );
+				if( stack != null )
+				{
+					stack.setStackSize( newSize );
+				}
+
+				if( diff != 0 && stack != null )
+				{
+					cachedStacks[slot] = newIS;
+					cachedAeStacks[slot] = stack;
+
+					final IAEItemStack a = stack.copy();
+					a.setStackSize( diff );
+					changes.add( a );
+				}
+			}
+		}
+
+		// Handle cases where the number of slots actually is lower now than before
+		if( slots < cachedStacks.length )
+		{
+			for (int slot = slots; slot < cachedStacks.length; slot++ )
+			{
+				IAEItemStack aeStack = cachedAeStacks[slot];
+				if ( aeStack != null ) {
+					IAEItemStack a = aeStack.copy();
+					a.setStackSize( -a.getStackSize() );
+					changes.add( a );
+				}
+			}
+
+			// Reduce the cache size
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		if( !changes.isEmpty() )
+		{
+			this.postDifference( changes );
+			return TickRateModulation.URGENT;
+		}
+		else
+		{
+			return TickRateModulation.SLOWER;
+		}
+	}
+
+	private boolean isDifferent( final ItemStack a, final ItemStack b )
+	{
+		if( a == b && b == null )
+		{
+			return false;
+		}
+
+		if( ( a == null && b != null ) || ( a != null && b == null ) )
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this condition so that it does not always evaluate to "true" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2583)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this condition so that it does not always evaluate to "true" [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2583)